### ISSUE TITLE
[REVIEW]: Heta compiler: a software tool for the development of large-scale QSP models and compilation into simulation formats

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3708]
**Submitting author:** @metelkin (<a href="http://orcid.org/0000-0001-6612-5373">Evgeny Metelkin</a>)
**Repository:** <a href="https://github.com/hetalang/heta-compiler" target ="_blank">https://github.com/hetalang/heta-compiler</a>
**Version:** v0.6.7
**Editor:** @fboehm
**Reviewer:** @martinmodrak, @elimillera
**Archive:** <a href="https://doi.org/10.5281/zenodo.5666487" target="_blank">10.5281/zenodo.5666487</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/ebff76c368d3adb720afe414ef6b29fb/status.svg)](https://joss.theoj.org/papers/ebff76c368d3adb720afe414ef6b29fb)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/ebff76c368d3adb720afe414ef6b29fb"><img src="https://joss.theoj.org/papers/ebff76c368d3adb720afe414ef6b29fb/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/ebff76c368d3adb720afe414ef6b29fb/status.svg)](https://joss.theoj.org/papers/ebff76c368d3adb720afe414ef6b29fb)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/hetalang/heta-compiler" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@martinmodrak & @elimillera, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @fboehm know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @martinmodrak
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/hetalang/heta-compiler">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/hetalang/heta-compiler">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@metelkin) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @elimillera
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/hetalang/heta-compiler">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/hetalang/heta-compiler">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@metelkin) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @martinmodrak, @elimillera it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3708 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.29 s (753.3 files/s, 85066.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                     133           1304           1708          11146
JSON                            54             73              0           7004
XML                             10            147             10           1871
Markdown                         8            307              0            892
YAML                            10             12              6            316
Julia                            2             30             26             83
R                                3             14             12             52
C++                              1             10              7             31
MATLAB                           1              0              0              7
-------------------------------------------------------------------------------
SUM:                           222           1897           1769          21402
-------------------------------------------------------------------------------


Statistical information for the repository 'c1caa7b1fcd3173aa381a9c4' was
gathered on 2021/09/09.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Evgeny Metelkin                705         35097          21395           98.34
metelkin                         5           897             54            1.66

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Evgeny Metelkin           14206           40.5         13.8               13.24
```

**Comment by fboehm** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: :point_left:

**Comment by fboehm** [comment]
@elimillera - Thanks again for agreeing to review! It looks like you might not have accepted the invitation from whedon, the JOSS bot. If this is the case, please feel free to accept the invitation so that you can check items in the above checklist. I think that the invitation expires in a few days. Thanks again!

**Comment by whedon** [comment]
:wave: @elimillera, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @martinmodrak, please update us on how your review is going (this is an automated reminder).

**Comment by elimillera** [comment]
Review is going well. I've reviewed the easy things to check and so far so good. Its been a couple years from when I was last working with models like these so I'm taking some time and checking some other sources as I go through it. I expect to be close to finished in the next 2-3 weeks.

**Comment by fboehm** [comment]
Thanks, @elimillera ! I appreciate your thoroughness. As you work through the review, please let me know if I can assist with anything. Thanks again!

**Comment by fboehm** [comment]
Thanks, @metelkin ! If you've addressed all of the issues from @elimillera , then there's nothing for you to do right now. we'll just need to wait for @martinmodrak to complete the review.

**Comment by martinmodrak** [comment]
Sorry to have kept everybody waiting. I finally managed to do a first pass through the software and paper. 

This is a neat piece of software that looks well documented and well written. So far I haven't encountered any major obstacle. 

My suggestions so far are minor, and may not even need immediate resolution for acceptance in JOSS.

1) The paper makes it look like Heta is the first tool with the ambition to build one modelling format that would serve many purposes. However, my (limited) understanding is that there are other contenders, notably SBML, which seems to have similar aims to Heta. I don't think I completely understand, what Heta adds that SMBL does not. If there are other 


2) I filed a single bug (https://github.com/hetalang/heta-compiler/issues/10), which I presume is something relatively benign and easy to fix. 

3) Possibly related is the fact that it appears (I didn't fully test) that the compiler is silent when an output format does not support a feature (e.g. not supporting some Switcher types). If that's the case, it might be helpful to raise a warning in such cases.

I still plan to spend some more time trying the software out, but wanted to share my results so far to let you react.

**Comment by metelkin** [comment]
@martinmodrak,
Thank you for your review.
1) The Heta language is not the first formal representation of biomodels. There are formats of storage like SBML, CellML, Antimony, as well as some DSLs that are applied in a specific modeling tools like mrgsolve, ModelingToolkit.jl, etc. Each of them has some specific purpose and limitations. There were two reasons not to mention them in the article. The main reason was the limitation in manuscript size. A lot of different aspects are skipped because of it. The another reason is that the Heta modeling formats is the secondary topic. The main goal is to present Heta-compiler as a modeling environment and its useful features. Anyway I guess the question that you mentioned is important and I am going to do the following: (i) I will include one or two sentences regarding the other formats (I need to check how many symbols I have); (ii) I will write the extended comparison in documentation site.
2) I have fixed it and replied to the issue you mentioned. The problem was the reserved word "default" in mrgsolve. You are right. Some warning would be helpful. We will do it in future releases. 
3) Actually Heta-compiler indicates errors and warnings if an export format does not support a feature or need attention. The Switcher component that you mentioned is the case. The compiler throws an error when exports to mrgsolve for unsupported features. But there are a lot of specific features (documented and undocumented) in each modeling format which we have to follow. I hope we will track them all in the future.

**Comment by martinmodrak** [comment]
> The main goal is to present Heta-compiler as a modeling environment and its useful features. Anyway I guess the question that you mentioned is important and I am going to do the following: (i) I will include one or two sentences regarding the other formats (I need to check how many symbols I have); (ii) I will write the extended comparison in documentation site.

I agree the focus is elsewhere. I would just find it helpful to learn what I gain by writing the model in Heta and then compiling to SBML as opposed to directly writing SBML. I am choosing SBML as it appears to have quite a broad support, so presumably many workflows with Heta would involve compiling to SBML to get access to the vast ecosystem of tools that support it. One or two sentences would probably be needed to satisfy the "State of the field" requirement, however no need for an extended comparison just for me. 

> Actually Heta-compiler indicates errors and warnings if an export format does not support a feature or need attention. The Switcher component that you mentioned is the case. 

Unfortunately, this doesn't appear to be the case: https://github.com/hetalang/heta-compiler/issues/11

There is also potentially problematic design in the way the definition of export targets is part of the model specification. One could IMHO argue that there should be a separation between model definition and export targets. E.g. if I want to share my model with the community, I probably don't want to prescribe how the community will export it. Conflicting needs of different team members could lead to unnecessary version control conflicts in the export definitionsm etc. The need for `--julia-only` and `--skip-export` CLI options shows that you are struggling a bit with this design choice already. 

Some (not mutually exclusive) options to consider: 

- Move export definitions to `platform.json` (presumably, I won't be sharing `platform.json` the same way I share model files?, i.e. two team members could share Heta code but have each their own platform?)
- Allow full control of exports from the CLI (e.g. allow stuff like `heta build --export "{ filepath: model, format: DBSolve }"`)
- Keep the current system, but encourage separating export definitions and models in two files (probably by the export file including the model), codify this as a good practice (e.g. by using it in examples)
- Have a make-like system, where the export definitions function as "targets" and one can specify `heta build [target-list]` to only build specific targets)

Generally, handling exports well seems like an important (but hard) design decision. As with the other stuff, I don't insist on this being resolved for acceptance in JOSS, just wanted to share my concerns.

In summary, I think the only really blocking problem is fulfilling the "State of the field" requirement by at least briefly contrasting what one can do with Heta that one could not accomplish with SBML or other common tool. There are a bunch of suggestions and two bugs I noticed, but I am OK with acceptance even if those are not addressed.

**Comment by metelkin** [comment]
> I would just find it helpful to learn what I gain by writing the model in Heta and then compiling to SBML as opposed to directly writing SBML. I am choosing SBML as it appears to have quite a broad support, so presumably many workflows with Heta would involve compiling to SBML to get access to the vast ecosystem of tools that support it. One or two sentences would probably be needed to satisfy the "State of the field" requirement, however no need for an extended comparison just for me.

I have added one sentence to mention the most known standard representation of bio-models. See the latest version in the "paper" branch. See below.

>The standardization of process-description modeling notation was also pursued in formats like SBML, CellML, Antimony. However the Heta standard can be distinguished by the specific features:  ... (list of Heta's features)

For a broader discussion here, I need to mention that SBML (as an example) and Heta are not competitors. Heta, in general, is a human-readable language that has been created to optimize the process of model development by users. SBML is an industrial xml-based standard for model exchange. I posted here a quote from SBML official site http://sbml.org
> SBML is a machine-readable format for representing models.
> What does SBML look like? Ugly. Don't look at it, unless you're developing software, in which case, you have to look at it, and we feel for you. SBML is really not meant to be edited by hand or exposed to users.
I know nobody who manually edits SBML in his everyday modeling workflow.

Anyway to develop the SBML model you have to use another layer: graphical programming (like in SimBiology), table creation (like in DBSolve), or use a DSL like in mrgsolve which is usually applicable for a single tool only.

**Comment by metelkin** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: :point_left:

**Comment by metelkin** [comment]
> Unfortunately, this doesn't appear to be the case: hetalang/heta-compiler#11
Fixed in master branch. It will be in release v0.6.7

**Comment by metelkin** [comment]
@martinmodrak, thanks again for such a deep analysis.
Regarding "export" design you are right. The current solution seems to be inconsistent in some way.
The "export" settings shouldn't be part of the standard and it is not part of a model at all.

Moreover, in one of the previous Heta versions "export" settings were part of the platfrom.json file. 
The decision to move \#export back to the main Heta code was made based on multiple users requests. 
It was easier for learning purposes.

I guess it is necessary to discuss the changes with endpoint users, i.e. modelers (some of them don't use GitHub at all).
It's problematic to modify this feature just now because many users use this option every day.

For now, I gonna create the issue and start this discussion there.

**Comment by martinmodrak** [comment]
Great. I think the software + paper now meets all criteria for acceptance in JOSS.

**Comment by metelkin** [comment]
Thank you, everyone!
@fboehm, what's next?

**Comment by fboehm** [comment]
Thanks, everyone! Let me proofread the paper before recommending it for publication. I might have a small list of typos that need to be fixed. I expect to get back to you, @metelkin, in the next few days.

**Comment by fboehm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #3708 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by fboehm** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: :point_left:

**Comment by fboehm** [comment]
Line 20: Define SB before using it as an abbreviation. Does it abbreviate "Systems biology"?
Line 115: Capitalize the "H" in "Heta". One approach is to put curly braces around the H in the bib file: {H}eta

**Comment by fboehm** [comment]
@metelkin - please make the two minor fixes that I suggested above.

**Comment by fboehm** [comment]
I can confirm that the DOIs in the bibliography all resolve to the intended articles.

**Comment by metelkin** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by metelkin** [comment]
@fboehm, I've updated the manuscript.

**Comment by fboehm** [comment]
Thank you, @metelkin!

**Comment by fboehm** [comment]
@whedon check references

**Comment by fboehm** [comment]
@metelkin - we now need you to make a tagged release and archive, and report the version number and archive DOI here in the review thread. 

Thanks again!

**Comment by metelkin** [comment]
Hello @fboehm 
I've published the release.

Version 0.6.7
10.5281/zenodo.5666487

**Comment by fboehm** [comment]
Thanks, @metelkin !

**Comment by fboehm** [comment]
@metelkin - I failed to tell you that the title of the archive needs to be the same as the title of the paper. Can you please fix this? Thanks again! Right now, it looks like the title of the archive is: hetalang/heta-compiler: Version 0.6.7 - ready for JOSS

**Comment by metelkin** [comment]
@fboehm - No problem. Done!

**Comment by fboehm** [comment]
@metelkin - Excellent!

**Comment by fboehm** [comment]
@whedon set v0.6.7 as version

**Comment by whedon** [comment]
OK. v0.6.7 is the version.

**Comment by fboehm** [comment]
set 10.5281/zenodo.5666487 as archive

**Comment by fboehm** [comment]
@whedon set 10.5281/zenodo.5666487 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5666487" target="_blank">10.5281/zenodo.5666487</a> is the archive.

**Comment by fboehm** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2741

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2741, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1093/bioinformatics/btp401 is OK
- 10.1093/bioinformatics/btg015 is OK
- 10.1002/psp4.12390 is OK
- 10.1002/cpt.1683 is OK
- 10.1007/s10441-018-9330-2 is OK
- 10.1016/j.csbj.2016.09.002 is OK
- 10.13140/RG.2.2.18555.13602 is OK
- 10.1007/978-1-60761-175-2_11 is OK

MISSING DOIs

- 10.1002/0471973165.ch23 may be a valid DOI for title: Cellular kinetic modeling of the microbial metabolism.

INVALID DOIs

- None
```

**Comment by fboehm** [comment]
Actually - @metelkin - can you add the doi suggested above to your bib file? I didn't see until now the output of the check references command. Thanks again!

**Comment by metelkin** [comment]
@fboehm - I've added the required doi to the bib file.

**Comment by fboehm** [comment]
Perfect! Thanks @metelkin !

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03708/joss.03708/10.21105.joss.03708.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1458587353262571522) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2743
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03708](https://doi.org/10.21105/joss.03708)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congratulations @metelkin on your article's publication in JOSS!

Many thanks to @martinmodrak and @elimillera for reviewing this, and @fboehm for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03708/status.svg)](https://doi.org/10.21105/joss.03708)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03708">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03708/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03708/status.svg
   :target: https://doi.org/10.21105/joss.03708
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03708/status.svg)](https://doi.org/10.21105/joss.03708)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
